src/mailkit.cs

```diff
@@ -0,0 +1,423 @@
+	}
+
+	[NoWatch, NoTV, NoiOS, NoMacCatalyst, Mac (12,0)]
+	[Protocol, Advice ("'MEContentBlocker' is not available in UIKit on macOS.")]
+	[BaseType (typeof (NSObject))]
+	interface MEContentBlocker
+	{
+		[Abstract]
+		[Export ("contentRulesJSON")]
+		NSData ContentRulesJson { get; }
+	}
+
+	[NoWatch, NoTV, NoiOS, NoMacCatalyst, Mac (12,0)]
+	[Protocol, Advice ("'MEExtension' is not available in UIKit on macOS.")]
+	[BaseType (typeof (NSObject))]
+	interface MEExtension
+	{
+		[Export ("handlerForComposeSession:")]
+		MEComposeSessionHandler HandlerForComposeSession (MEComposeSession session);
+
+		[Export ("handlerForMessageActions")]
+		MEMessageActionHandler HandlerForMessageActions { get; }
+
+		[Export ("handlerForContentBlocker")]
+		MEContentBlocker HandlerForContentBlocker { get; }
+
+		[Export ("handlerForMessageSecurity")]
+		IMEMessageSecurityHandler HandlerForMessageSecurity { get; }
+	}
+
+	[NoWatch, NoTV, NoiOS, NoMacCatalyst, Mac (12,0)]
+	[Protocol, Advice ("'MEMessageActionHandler' is not available in UIKit on macOS.")]
+	[BaseType (typeof (NSObject))]
+	interface MEMessageActionHandler
+	{
+		[Abstract]
+		[Export ("decideActionForMessage:completionHandler:")]
+		void CompletionHandler (MEMessage message, Action<MEMessageActionDecision> completionHandler);
+
+		[Export ("requiredHeaders", ArgumentSemantic.Copy)]
+		string[] RequiredHeaders { get; }
+	}
+
+	[NoWatch, NoTV, NoiOS, NoMacCatalyst, Mac (12,0)]
+	[Protocol, Advice ("'MEMessageDecoder' is not available in UIKit on macOS.")]
+	[BaseType (typeof (NSObject))]
+	interface MEMessageDecoder
+	{
+		[Abstract]
+		[Export ("decodedMessageForMessageData:")]
+		[return: NullAllowed]
+		MEDecodedMessage DecodedMessage (NSData data);
+	}
+
+	[NoWatch, NoTV, NoiOS, NoMacCatalyst, Mac (12,0)]
+	[Protocol, Advice ("'MEMessageEncoder' is not available in UIKit on macOS.")]
+	[BaseType (typeof (NSObject))]
+	interface MEMessageEncoder
+	{
+		[Abstract]
+		[Export ("getEncodingStatusForMessage:completionHandler:")]
+		void GetEncodingStatus (MEMessage message, Action<MEOutgoingMessageEncodingStatus> completionHandler);
+
+		[Abstract]
+		[Export ("encodeMessage:shouldSign:shouldEncrypt:completionHandler:")]
+		void EncodeMessage (MEMessage message, bool shouldSign, bool shouldEncrypt, Action<MEMessageEncodingResult> completionHandler);
+	}
+
+	interface IMEMessageSecurityHandler {}
+
+	[NoWatch, NoTV, NoiOS, NoMacCatalyst, Mac (12,0)]
+	[Protocol, Advice ("'MEMessageSecurityHandler' is not available in UIKit on macOS.")]
+	interface MEMessageSecurityHandler : MEMessageEncoder, MEMessageDecoder
+	{
+		[Abstract]
+		[Export ("extensionViewControllerForMessageSigners:")]
+		[return: NullAllowed]
+		MEExtensionViewController ExtensionViewController (MEMessageSigner[] messageSigners);
+	}
+
+	[NoWatch, NoTV, NoiOS, NoMacCatalyst, Mac (12,0)]
+	[DisableDefaultCtor]
+	[BaseType (typeof (NSViewController))]
+	interface MEExtensionViewController
```

Comment:
Introspection on Mac gives this error:
```
14:36:15.5618610 	[FAIL] DesignatedInitializer :   1 potential errors found in 751 designated initializer validated
14:36:15.5619810   Expected: 0
14:36:15.5619930   But was:  1
14:36:15.5620130 
14:36:15.5620240 [FAIL] MailKit.MEExtensionViewController should re-expose NSViewController::.ctor(System.String, Foundation.NSBundle)
14:36:15.5620320 
14:36:15.5624080 		  at Introspection.ApiCtorInitTest.DesignatedInitializer () [0x0014b] in /Users/tjlambert/Documents/Xcode13/xamarin-macios/tests/introspection/ApiCtorInitTest.cs:335 

```

This seems like something that should be ignored in the test?

Reply:
From the documentation about [MEExtensionViewController](https://developer.apple.com/documentation/mailkit/meextensionviewcontroller?language=objc), app developers are supposed to subclass this class to provide their own implementation. This means that there must be constructors available in this class for subclasses to call. In Objective-C constructors are automatically available to all subclasses (not only direct ones), which means that the constructors defined on `NSViewController` are available for this class as well (that's what the failing test is talking about).

In the end, it seems like we should add two constructors: the default one, and the one from the error message, and make both protected (which effectively makes this class `abstract`, because you can only instantiate it if you subclass it - which is what the documentation implies should be the case):

```csharp
[NoWatch, NoTV, NoiOS, NoMacCatalyst, Mac (12,0)]
[DefaultCtorVisibility (Visibility.Protected)]
[BaseType (typeof (NSViewController))]
interface MEExtensionViewController
{
		[DesignatedInitializer]
		[Protected]
		[Export ("initWithNibName:bundle:")]
		IntPtr Constructor ([NullAllowed] string nibNameOrNull, [NullAllowed] NSBundle nibBundleOrNull);
}
```

Reply:
Thank you! I do (now) recall Rachel asking about this and I will try to remember this for next time!